ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs

```diff
@@ -0,0 +1,364 @@
+           |> not
+
+    let rec private isFullyAnnotatedFunctionTypeUsage (functionTypeUsage: IFunctionTypeUsage) =
+        isFullyAnnotatedTypeUsage functionTypeUsage.ArgumentTypeUsage &&
+        isFullyAnnotatedTypeUsage functionTypeUsage.ReturnTypeUsage
+
+    and isFullyAnnotatedTypeUsage (typeUsage: ITypeUsage) =
+        match typeUsage with
+        | :? INamedTypeUsage as namedTypeUsage ->
+            isFullyAnnotatedNamedTypeUsage namedTypeUsage
+        | :? IFunctionTypeUsage as functionTypeUsage ->
+            isFullyAnnotatedFunctionTypeUsage functionTypeUsage
+        | :? IConstrainedTypeUsage as constrainedTypeUsage ->
+            isFullyAnnotatedTypeUsage constrainedTypeUsage.TypeUsage
+        | _ -> false
+
+    let isFullyAnnotatedReturnTypeInfo (returnTypeInfo: IReturnTypeInfo) =
+        isNotNull returnTypeInfo.ReturnType &&
+        isFullyAnnotatedTypeUsage returnTypeInfo.ReturnType
+
+    let rec isFullyAnnotatedPattern (pattern: IFSharpPattern) =
+        match pattern.IgnoreInnerParens() with
+        | :? IUnitPat ->
+            true
+        | :? ITypedPat as typedPat ->
+            isFullyAnnotatedPattern typedPat.Pattern
+        | pattern ->
+            pattern |> PatUtil.getTypeUsage |> ValueOption.exists isFullyAnnotatedTypeUsage
+            || pattern |> PatUtil.getReturnTypeInfo |> ValueOption.exists isFullyAnnotatedReturnTypeInfo
+
+    let isFullyAnnotatedBinding (binding: IBinding) =
+        isNotNull binding.ReturnTypeInfo &&
+        binding.ParametersDeclarationsEnumerable
+        |> Seq.forall (fun parameter ->
+            isFullyAnnotatedPattern parameter.Pattern)
+
+    let isFullyAnnotatedMemberDeclaration (memberDeclaration: IMemberDeclaration) =
+        isNotNull memberDeclaration.ReturnTypeInfo &&
+        isFullyAnnotatedReturnTypeInfo memberDeclaration.ReturnTypeInfo &&
+        memberDeclaration.ParametersDeclarations
+        |> Seq.forall (fun parameter ->
+            isFullyAnnotatedPattern parameter.Pattern)
+
+    // let f x ... =
+    // let f<'a> = ...
+    let isFunctionBinding (binding: IBinding) =
+        binding.HasParameters
+        || binding.HeadPattern.GetNextMeaningfulSibling() :? IPostfixTypeParameterDeclarationList
+
+    // let x = y
+    // let x = fun y -> ...
+    let isValueBinding (binding: IBinding) =
+       not (isFunctionBinding binding)
+
+module SpecifyUtil =
+
+    let private addSpaceBeforeColon forceSpaceBeforeColon (pattern: ITreeNode) =
+        if forceSpaceBeforeColon || pattern.GetSettingsStoreWithEditorConfig().GetValue(fun (key: FSharpFormatSettingsKey) -> key.SpaceBeforeColon) then
+            ModificationUtil.AddChildBefore(pattern, Whitespace()) |> ignore
+
+    let private addReturnTypeInfo typeString (node: ITreeNode) =
+        let factory = node.CreateElementFactory()
+        let typeUsage = factory.CreateTypeUsage(typeString)
+        ModificationUtil.AddChildAfter(node, factory.CreateReturnTypeInfo(typeUsage))
```

Comment:
I wonder if using generated `IBindingLikeDeclaration.SetReturnTypeInfo` would produce the expected result here instead?

Even if it doesn't work for some reason, we shouldn't probably add a function that can add a return type info to any node type.

Reply:
We found that this api most probably doesn't work now. I've removed this function and inlined implementations into specifyBinding/Method/PropReturnType